File: frb_example/pure_dart/lib/src/rust/api/chrono_type.dart
// ignore_for_file: invalid_use_of_internal_member, unused_import

import '../frb_generated.dart';
import 'package:flutter_rust_bridge/flutter_rust_bridge_for_generated.dart';

Future<DateTime> datetimeUtc({required DateTime d, dynamic hint}) =>
    RustLib.instance.api.datetimeUtc(d: d, hint: hint);

Future<DateTime> datetimeLocal({required DateTime d, dynamic hint}) =>
    RustLib.instance.api.datetimeLocal(d: d, hint: hint);

Future<DateTime> naivedatetime({required DateTime d, dynamic hint}) =>
    RustLib.instance.api.naivedatetime(d: d, hint: hint);

Future<DateTime?> optionalEmptyDatetimeUtc({DateTime? d, dynamic hint}) =>
    RustLib.instance.api.optionalEmptyDatetimeUtc(d: d, hint: hint);

Future<Duration> duration({required Duration d, dynamic hint}) =>
    RustLib.instance.api.duration(d: d, hint: hint);

Future<List<Duration>> handleTimestamps(
        {required List<DateTime> timestamps,
        required DateTime epoch,
        dynamic hint}) =>
    RustLib.instance.api
        .handleTimestamps(timestamps: timestamps, epoch: epoch, hint: hint);

Future<List<DateTime>> handleDurations(
        {required List<Duration> durations,
        required DateTime since,
        dynamic hint}) =>
    RustLib.instance.api
        .handleDurations(durations: durations, since: since, hint: hint);

Future<TestChrono> testChrono({dynamic hint}) =>
    RustLib.instance.api.testChrono(hint: hint);

Future<TestChrono> testPreciseChrono({dynamic hint}) =>
    RustLib.instance.api.testPreciseChrono(hint: hint);

Future<Duration> howLongDoesItTake(
        {required FeatureChrono mine, dynamic hint}) =>
    RustLib.instance.api.howLongDoesItTake(mine: mine, hint: hint);

class FeatureChrono {
  final DateTime utc;
  final DateTime local;
  final Duration duration;
  final DateTime naive;

  const FeatureChrono({
    required this.utc,
    required this.local,
    required this.duration,
    required this.naive,
  });

  @override
  int get hashCode =>
      utc.hashCode ^ local.hashCode ^ duration.hashCode ^ naive.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is FeatureChrono &&
          runtimeType == other.runtimeType &&
          utc == other.utc &&
          local == other.local &&
          duration == other.duration &&
          naive == other.naive;
}

class TestChrono {
  final DateTime? dt;
  final DateTime? dt2;
  final Duration? du;

  const TestChrono({
    this.dt,
    this.dt2,
    this.du,
  });

  @override
  int get hashCode => dt.hashCode ^ dt2.hashCode ^ du.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is TestChrono &&
          runtimeType == other.runtimeType &&
          dt == other.dt &&
          dt2 == other.dt2 &&
          du == other.du;
}

File: frb_example/pure_dart/rust/src/api/chrono_type.rs
pub fn datetime_utc_twin_normal(d: chrono::DateTime<chrono::Utc>) -> chrono::DateTime<chrono::Utc> {
    use chrono::Datelike;
    use chrono::Timelike;
    assert_eq!(&d.year(), &2022);
    assert_eq!(&d.month(), &9);
    assert_eq!(&d.day(), &10);
    assert_eq!(&d.hour(), &20);
    assert_eq!(&d.minute(), &48);
    assert_eq!(&d.second(), &53);
    #[cfg(target_arch = "wasm32")]
    assert_eq!(&d.nanosecond(), &123_000_000);
    #[cfg(not(target_arch = "wasm32"))]
    assert_eq!(&d.nanosecond(), &123_456_000);
    d
}

pub fn datetime_local_twin_normal(
    d: chrono::DateTime<chrono::Local>,
) -> chrono::DateTime<chrono::Local> {
    use chrono::Datelike;
    use chrono::Timelike;
    assert_eq!(&d.year(), &2022);
    assert_eq!(&d.month(), &9);
    assert_eq!(&d.day(), &10);
    if cfg!(target_arch = "wasm32") {
        assert_eq!(&d.nanosecond(), &123_000_000);
    } else {
        assert_eq!(&d.hour(), &20);
        assert_eq!(&d.nanosecond(), &123_456_000);
    }
    assert_eq!(&d.minute(), &48);
    assert_eq!(&d.second(), &53);
    d
}

pub fn naivedatetime_twin_normal(d: chrono::NaiveDateTime) -> chrono::NaiveDateTime {
    use chrono::{Datelike, Timelike};
    assert_eq!(&d.year(), &2022);
    assert_eq!(&d.month(), &9);
    assert_eq!(&d.day(), &10);
    assert_eq!(&d.hour(), &20);
    assert_eq!(&d.minute(), &48);
    assert_eq!(&d.second(), &53);
    #[cfg(target_arch = "wasm32")]
    assert_eq!(&d.nanosecond(), &123_000_000);
    #[cfg(not(target_arch = "wasm32"))]
    assert_eq!(&d.nanosecond(), &123_456_000);
    d
}

pub fn optional_empty_datetime_utc_twin_normal(
    d: Option<chrono::DateTime<chrono::Utc>>,
) -> Option<chrono::DateTime<chrono::Utc>> {
    assert_eq!(&d, &None);
    d
}

pub fn duration_twin_normal(d: chrono::Duration) -> chrono::Duration {
    assert_eq!(&d.num_hours(), &4);
    d
}

pub fn handle_timestamps_twin_normal(
    timestamps: Vec<chrono::NaiveDateTime>,
    epoch: chrono::NaiveDateTime,
) -> Vec<chrono::Duration> {
    timestamps
        .into_iter()
        .map(|ts| epoch.signed_duration_since(ts))
        .collect()
}

pub fn handle_durations_twin_normal(
    durations: Vec<chrono::Duration>,
    since: chrono::DateTime<chrono::Local>,
) -> Vec<chrono::DateTime<chrono::Local>> {
    durations.into_iter().map(|dur| since - dur).collect()
}

pub struct TestChronoTwinNormal {
    pub dt: Option<chrono::DateTime<chrono::Utc>>,
    pub dt2: Option<chrono::NaiveDateTime>,
    pub du: Option<chrono::Duration>,
}

pub fn test_chrono_twin_normal() -> TestChronoTwinNormal {
    TestChronoTwinNormal {
        dt: Some(chrono::DateTime::from_naive_utc_and_offset(
            chrono::DateTime::from_timestamp(1631297333, 0)
                .unwrap()
                .naive_utc(),
            chrono::Utc,
        )),
        dt2: Some(
            chrono::DateTime::from_timestamp(1631297333, 0)
                .unwrap()
                .naive_utc(),
        ),
        du: Some(chrono::Duration::hours(4)),
    }
}

pub fn test_precise_chrono_twin_normal() -> TestChronoTwinNormal {
    TestChronoTwinNormal {
        dt: Some(chrono::DateTime::from_naive_utc_and_offset(
            chrono::DateTime::from_timestamp(1014466435, 0)
                .unwrap()
                .naive_utc(),
            chrono::Utc,
        )),
        dt2: Some(
            chrono::DateTime::from_timestamp(-5362715015, 0)
                .unwrap()
                .naive_utc(),
        ),
        du: Some(chrono::Duration::hours(4)),
    }
}

#[derive(Debug, Clone)]
pub struct FeatureChronoTwinNormal {
    pub utc: chrono::DateTime<chrono::Utc>,
    pub local: chrono::DateTime<chrono::Local>,
    pub duration: chrono::Duration,
    pub naive: chrono::NaiveDateTime,
}

pub fn how_long_does_it_take_twin_normal(
    mine: FeatureChronoTwinNormal,
) -> anyhow::Result<chrono::Duration> {
    use chrono::{Datelike, Timelike};
    let difference: chrono::Duration = chrono::Utc::now() - mine.utc;
    assert_eq!(&mine.duration.num_hours(), &4);
    assert_eq!(&mine.naive.year(), &2022);
    assert_eq!(&mine.naive.month(), &9);
    assert_eq!(&mine.naive.day(), &10);
    assert_eq!(&mine.naive.hour(), &20);
    assert_eq!(&mine.naive.minute(), &48);
    assert_eq!(&mine.naive.second(), &53);
    #[cfg(target_arch = "wasm32")]
    assert_eq!(&mine.naive.nanosecond(), &123_000_000);
    #[cfg(not(target_arch = "wasm32"))]
    assert_eq!(&mine.naive.nanosecond(), &123_456_000);
    Ok(difference)
}
